pkg/segment/query/processor/searcher.go

```diff
@@ -387,13 +424,66 @@ func (s *Searcher) initializeQSRs() error {
 	return nil
 }
 
-func (s *Searcher) getBlocks() ([]*block, error) {
-	qsrs, err := query.GetSortedQSRs(s.queryInfo, s.startTime, s.querySummary)
-	if err != nil {
-		log.Errorf("qid=%v, searcher.getBlocks: failed to get sorted QSRs: %v", s.qid, err)
-		return nil, err
+func (s *Searcher) getFilteredBlocks(blocks []*block) []*block {
+	filteredBlocks := make([]*block, 0)
+
+	for _, block := range blocks {
+		if s.processedBlocks[block.parentQSR.GetSegKey()] == nil {
+			s.processedBlocks[block.parentQSR.GetSegKey()] = make(map[uint16]struct{})
+		}
+		_, processed := s.processedBlocks[block.parentQSR.GetSegKey()][block.BlkNum]
+		if processed {
+			continue
+		}
+		if block.HighTs >= s.minTimestamp {
+			filteredBlocks = append(filteredBlocks, block)
+			s.processedBlocks[block.parentQSR.GetSegKey()][block.BlkNum] = struct{}{}
+		}
+	}
+
+	return filteredBlocks
+}
+
+func (s *Searcher) getQSRSToProcess() []*query.QuerySegmentRequest {
+	qsrs := make([]*query.QuerySegmentRequest, 0)
+
+	prevMinTimestamp := s.minTimestamp
+
+	s.segmentIdx += segmentBatchOffset
+	if s.segmentIdx >= len(s.qsrs) {
+		s.segmentIdx = len(s.qsrs)
+		s.minTimestamp = 0
+		s.gotAllSegments = true
+	} else {
+		s.minTimestamp = s.qsrs[s.segmentIdx].GetStartEpochMs()
 	}
 
+	// Batch all the segments having records within the startTime and are not yet processed
+	for i := 0; i < len(s.qsrs); i++ {
+		_, processed := s.processedSegment[s.qsrs[i].GetSegKey()]
+		if processed {
+			continue
```

Comment:
nit: it might be slightly more performant to have a list of unprocessed QSRs instead of a list of all QSRs and a map of which QSRs have already been processed. Is there a reason you did it this way?

Reply:
addressed in the latest commit.